# 4. 에러&예외 처리 (error_exception)

## 문법 에러(Syntax Error)

> 문법 에러가 있는 프로그램은 실행되지 않습니다.

## 예외(Exception)

> 실행 도중 예상하지 못한 상황(exception)을 맞이하면, 프로그램 실행을 멈춥니다.

ZeroDivisionError  어떤 수를 0으로 나누게 되면 에러가 발생

NameError  			어느 곳에서도 정의되지 않은 변수를 호출 하였을 경우 에러가 발생

TypeError				자료형이 올바르지 못한 경우에도 에러가 발생

ValueError			   존재하지 않는 **값**을 찾고자 할 경우 에러가 발생

IndexError 			  존재하지 않는 **index로 조회**할 경우 에러가 발생

KeyError				   딕셔너리는 존재하지 않는 Key로 접근하게 되면 에러가 발생

ModuleNotFoundError 모듈을 찾을 수 없는 경우 에러가 발생

ImportError		 	모듈을 찾았으나 가져오는 과정에서 실패하는 경우 에러가 발생

KeyboardInterrupt 무한 반복되는 while문을 실행시켜보고, 정지시킬경우 에러가 발생



# 예외 처리(Exception Handling)

## `try` & `except`

`try` 문을 이용하여 예외 처리를 할 수 있습니다.

------

### 기초 문법

```python
try:
    <코드 블럭 1>
except (예외):
    <코드 블럭 2>
```

- `try` 아래의 코드블락(code block)이 실행됩니다.
- 예외가 발생되지 않으면, **`except`없이 실행이 종료 됩니다.**
- 예외가 발생하면, **남은 부분을 수행하지 않고**, `except`가 실행됩니다.



### 복수의 예외 처리

하나 이상의 예외를 모두 처리할 수 있습니다.

괄호가 있는 튜플로 여러 개의 예외를 지정할 수 있습니다.

------

**활용법**

```python
try:
    <코드 블럭 1>
except (예외1, 예외2):
    <코드 블럭 2>


try:
    <코드 블럭 1>
except 예외1:
    <코드 블럭 2>
except 예외2:
    <코드 블럭 3>
```



### `else`

- 에러가 발생하지 않는 경우 수행되는 문장은 `else`를 이용합니다.
- 모든 except 절 뒤에와야 합니다.
- try 절이 예외를 일으키지 않을 때 실행되어야만 하는 코드에 적절합니다.

------

**활용법**

```python
try:
    <코드 블럭 1>
except 예외:
    <코드 블럭 2>
else:
    <코드 블럭 3>
```



### `finally`

- 반드시 수행해야하는 문장은 `finally`를 활용합니다.
- 즉, 모든 상황에 실행되어야만 하는 코드를 정의하는데 활용합니다.
- 예외의 발생 여부과 관계없이 try 문을 떠날 때 항상 실행합니다.

------

**활용법**

```python
try:
    <코드 블럭 1>
except 예외:
    <코드 블럭 2>
finally:
    <코드 블럭 3>
```



### 에러 메시지 처리 `as`

`as` 키워드를 활용하여 에러 메시지를 보여줄 수도 있습니다.

------

**활용법**

```python
try:
    <코드 블럭 1>
except 예외 as err:
    <코드 블럭 2>
```



## 예외 발생 시키기(Exception Raising)

### `raise`

`raise`를 통해 예외를 강제로 발생시킬 수 있습니다.

------

**활용법**

```python
raise <에러>('메시지')
```